### ISSUE TITLE
Connection unit-testing

### ISSUE THREAD

**Issue by palkan** [palkan/action-cable-testing#1]
There is no way to test `ApplicationCable::Connection` yet.

Why do we need? First, to test our authentication and identifiers resolving (in `#connect` method), secondly, to test our `#disconnect` code.

The tricky moment here is that we have to deal with HTTP requests somehow (like in integration tests), but we don't want to run the whole Action Cable server and use real WebSocket clients.

We can extend our `ActionCable::Channel::TestCase` with the following API:

```ruby
def test_connect
  # What do we need? Connection path ('cause it may contain query params), headers and cookies
  connect path: "/cable?some_param=val", cookies: { user_id: "1123" }, headers: { "X-API-TOKEN" => 'abc' }
  
  assert connection.connected?
end

def test_connection_rejected
  assert_raises ActionCable::Connection::Authorization::UnauthorizedError do
    connect headers: { "X-API-TOKEN" => 'wrong_token' }
  end
end

def test_disconnect
  stub_connection(user_id: 123)

  assert_broadcast_on("users", event: "user_leave", id: "213) do
    disconnect
  end
end
```

**Comment by christianrolle** [comment]
@palkan Awesome! Testing the connection is the missing piece for my test coverage.
How can help out to push things forward (but I guess I only can maker smaller contributions)?

**Comment by palkan** [comment]
Sorry, I forgot about it(

> How can I help out to push things forward

Just ping me from time to time)

I hope to find a few hours this weekend to finish it

**Comment by ZackMattor** [comment]
Awesome! Also looking forward to this. Currently we're doing the following, so this API will be a welcome change!

Currently we require a header present for API authorization for the connection... so we have a line like `if request.headers["X-Api-Key"].present?` but this is throwing ` expected ActionCable::Connection::Authorization::UnauthorizedError, got #<TypeError: no implicit conversion of ActiveSupport::StringInquirer into Hash> with backtrace:` during tests. I'm going to try mocking this out so request returns a double... but do you know why this only happens in tests?

```ruby
RSpec.describe ApplicationCable::Connection do
  before :each do
    @server = ActionCable::Server::Base.new
    @env = Rails.env
  end

  describe 'connect' do
    it 'allows logged in users to subscribe' do
      user = create(:user)
      cookies = MockCookies.new(user_id: user.id)
      connection = ApplicationCable::Connection.new(@server, @env)
      expect(connection).to receive(:cookies).and_return(cookies)
      connection.connect
      expect(connection.current_user).to eq(user)
    end

    it 'disallows subscribing by unlogged users' do
      cookies = MockCookies.new({})
      connection = ApplicationCable::Connection.new(@server, @env)
      expect(connection).to receive(:cookies).and_return(cookies)
      expect { connection.connect }.to raise_error(ActionCable::Connection::Authorization::UnauthorizedError)
      expect(connection.current_user).to eq(nil)
    end
  end
end
```

**Comment by palkan** [comment]
Closed and released as 0.3.0. See release notes: https://github.com/palkan/action-cable-testing/releases/tag/v0.3.0
